docs/adr/2023-08-18_025-resource-based-api.md

```diff
@@ -0,0 +1,146 @@
+
+  - Need to poll `GetUTxO` _or_ aggregate confirmed transactions on client side
+    to know the latest UTxO for constructing transactions.
+
+  - Inclusion of the whole UTxO in the head is not always desirable and
+    filtering by address would be beneficial. (not addressed in this ADR though)
+
+  - As [ADR-15](/adr/15) also proposes, some clients may not need (or should
+    not have) access to administrative information.
+
+- It is often a good idea to separate the responsibilities of Commands and
+  Queries (CQRS), as well as the model they use.
+
+## Decision
+
+- Do not add a general purpose querying interface (e.g. graphql) to the
+  `hydra-node` and stay true to [ADR-3](/adr/3) of using duplex communication
+  channels (using websockets).
+
+- **Drop** `GetUTxO` and `GetUTxOResponse` messages as they advocate a
+  request/response way of querying.
+
+- Realize that `ClientInput` data is actually a `ClientCommand` (renaming them)
+  and that `ServerOutput` are just `projections` of the [internal event stream
+  (see ADR-24)](/adr/24) into read `models` on the API layer.
+
+- Compose the API out of resource `models`, which compartmentalize the domain
+  into topics on the API layer.
+
+  - Each resource has a _latest_ state, which can be queried or clients may
+    subscribe to any model changes.
+
+  - A resource's `model` type needs to be a result of a pure `projection` from
+    server output events, i.e. `project :: model -> ServerOutput -> model`.
+
+    - TBD: or use the event stream type `StateChanged` instead? potentially
+      renaming it?
+
+  - Each resource is available at some HTTP path, also called "endpoint":
+
+    - `GET` requests must respond with the _latest_ state in a single response.
+
+    - `GET` requests with websocket upgrade headers must start a websocket
+      connection, push the _latest_ state as first message and any resource
+      state updates after.
+
+    - Other HTTP verbs may be accepted by a resource handler, i.e. to issue
+      resource-specific commands. Any commands accepted must also be available
+      via the corresponding websocket connection.
+
+  - Each resource is only available as `JSON` encoding (for now).
+
+    - (TBD): already specify that `Accept` headers should be used to negotiate a
+      `content-type`?
+
+- (TBD): Keep the direct/raw output of `ServerOutput` events on `/`, which
+  also accepts all `ClientCommand` messages.
+
+### Resources
+
+Resource paths + HTTP verbs mapped to existing things:
+
+| Path                        | GET                         | POST                   | PATCH   | DELETE             |
+| :-------------------------- | :-------------------------- | :--------------------- | ------- | :----------------- |
+| `/node/head/status`         | `HeadStatus(..)`            | -                      | -       | -                  |
+| `/node/head/utxo`           | confirmed utxo              | -                      | -       | -                  |
```

Comment:
I'm guessing _utxo_ is meant for _utxo set_ here.

If I'm correct, what would mean _confirmed utxo_? Is it for the utxo that are included in a snapshot?

Also, I've noticed we currently have some issues with the clients only having access to the utxo included in a snapshot (see #612 ). I feel like `/node/head/utxo`, if containing the set of utxo included in a snapshot would exhibit the same problem and I would like us to solve it before. In my opinion, the client should, by default, only access the latest view of its node ledger when constructing a transaction. Whether or not these utxo have been included in a snapshot should not be considered at this step.

Last, but not least, we can currently have issues when the utxo set is very big and the _snapshot confirmed_ message is heavy and slows down the node. That led some users to ask for utxo query primitives instead of getting the full utxo set. I'm mentioning this, although I'm not sure we want to address it right now as having a huge utxo set would make it impossible to close the head with the current implementation anyway.

Reply:
I agree: We should probably already provide sub-paths for individual UTxO and we already have had requests for more powerful UTxO query capabilities (eg. https://github.com/input-output-hk/hydra/discussions/797)

Reply:
Good points.

We usually mean any number of unspent transaction outputs when we just say `utxo` or `UTxO`. This is also how it is documented right now in the api reference and glossary IIRC.

I commented below that we likely need both, local and confirmed versions of utxo and transactions.

I will add a section on how the querying of such big resources could be improved using content types and query parameters (and maybe even pagination?).